ARROW-17825: [C++] Allow the possibility to write several tables in ORCFileWriter

LouisClt opened the pull request:
I had the need to write an ORC file little by little, so as to not consume too much memory.
Following [this](https://github.com/apache/arrow/issues/14211) discussion, it appeared that the API did not seemed to prevent doing that, but that the internal implementation was not reusing the writer accordingly.

This PR makes the needed changes to reuse the "writer_" correctly.

I do not think that the preceding behaviour was correct, as calling several time the "Write" method would lead to incorrect ORC files.

LouisClt commented:
Yes I will do this once I made sure that the existing tests still pass.

wjones127 commented on cpp/src/arrow/adapters/orc/adapter_test.cc:
```diff
@@ -730,4 +733,30 @@ TEST_F(TestORCWriterSingleArray, WriteListOfMap) {
+  random::RandomArrayGenerator rand;
+};
+
+TEST_F(TestORCWriterMultipleWrite, MultipleWritesIntField) {
```
Nice test!

Could you also add one to verify that we get a sensible error if you write a table but then try to write a second table with a different schema?

LouisClt replied:
Yes, I can try do do this. Though, I will need to modify the code I think.

LouisClt replied:
Done

wjones127 commented on cpp/src/arrow/adapters/orc/adapter.h:
```diff
@@ -272,9 +272,9 @@ class ARROW_EXPORT ORCFileWriter {
 
-  /// \brief Write a table
+  /// \brief Write a table. This can be called multiple times.
   ///
```
```suggestion
  ///
  /// Tables passed in subsequent calls must match the schema of the table that was written first.
  ///
```

LouisClt commented:
I added a test for checking the schema. I added a specific error, and checked that an error was found. (however I didn't test that the error was the specific one ).

wjones127 commented on cpp/src/arrow/adapters/orc/adapter_test.cc:
```diff
@@ -730,4 +735,55 @@ TEST_F(TestORCWriterSingleArray, WriteListOfMap) {
+  auto array_int2 = rand.ArrayOf(int64(), num_rows, 0);
+  std::shared_ptr<Schema> input_schema2 = schema({field("col0", array_int2->type())});
+
+  auto input_chunked_array = std::make_shared<ChunkedArray>(array_int),
+       input_chunked_array2 = std::make_shared<ChunkedArray>(array_int2);
+  std::shared_ptr<Table> input_table = Table::Make(input_schema, {input_chunked_array}),
+                         input_table2 =
+                             Table::Make(input_schema2, {input_chunked_array2});
```
There is an overload of `Table::Make()` that takes Arrays, so we can just use that. Also, we generally repeat the type if we are declaring and assigning.

```suggestion
  std::shared_ptr<Table> input_table = Table::Make(input_schema, {array_int});
  std::shared_ptr<Table> input_table2 = Table::Make(input_schema2, {array_int2});
```

LouisClt replied:
Ok, thanks. I adapted the code from the "AssertArrayWriteReadEqual" method, so I kept a good part of the syntax

wjones127 commented on cpp/src/arrow/adapters/orc/adapter_test.cc:
```diff
@@ -730,4 +735,55 @@ TEST_F(TestORCWriterSingleArray, WriteListOfMap) {
+
+TEST_F(TestORCWriterMultipleWrite, MultipleWritesIntField) {
+  const int64_t num_rows = 1234;
+  const int nb_writes = 5;
```
We generally use `num_` instead of `nb_` in Arrow, so consider changing this.

LouisClt replied:
Ok, no problem. I will change this to num_writes.

wjones127 commented on cpp/src/arrow/adapters/orc/adapter_test.cc:
```diff
@@ -730,4 +735,55 @@ TEST_F(TestORCWriterSingleArray, WriteListOfMap) {
+  ASSERT_NOT_OK(writer->Write(*input_table2));
+
+  ARROW_EXPECT_OK(writer->Close());
+  EXPECT_OK_AND_ASSIGN(auto buffer, buffer_output_stream->Finish());
```
Were you planning on doing something with this buffer?

LouisClt replied:
Indeed, there is no need of this here

wjones127 commented on cpp/src/arrow/adapters/orc/adapter_test.cc:
```diff
@@ -239,7 +241,9 @@ void AssertTableWriteReadEqual(const std::shared_ptr<Table>& input_table,
   EXPECT_OK_AND_ASSIGN(auto writer, adapters::orc::ORCFileWriter::Open(
                                         buffer_output_stream.get(), write_options));
-  ARROW_EXPECT_OK(writer->Write(*input_table));
+  for (int iWrite = 0; iWrite < nb_writes; iWrite++) {
```
We take the convention from the Google Style Guide that variable names should be lower case.

```suggestion
  for (int i_write = 0; i_write < nb_writes; i_write++) {
```

wjones127 commented on cpp/src/arrow/adapters/orc/adapter_test.cc:
```diff
@@ -730,4 +735,55 @@ TEST_F(TestORCWriterSingleArray, WriteListOfMap) {
+  ARROW_EXPECT_OK(writer->Write(*input_table));
+
+  // This should not pass
+  ASSERT_NOT_OK(writer->Write(*input_table2));
```
How about this?
```suggestion
  ASSERT_RAISES(Status::Invalid, writer->Write(*input_table2));
```

LouisClt replied:
It will be better indeed.

wjones127 reviewed (changes requested):
This is looking good. Thanks for implementing this.

I have a few minor suggestions before it's fully ready.

wjones127 reviewed (approved):
This looks good to me. @pitrou would you be willing to take a look and merge if it looks good to you?

pitrou commented on cpp/src/arrow/adapters/orc/adapter.cc:
```diff
@@ -727,12 +727,22 @@ class ORCFileWriter::Impl {
+    } else {
+      bool schemas_matching = table.schema()->Equals(arrow_schema_, false);
+      if (!schemas_matching) {
+        return (Status(
+            StatusCode::Invalid,
```
Can use shortcut methods instead
```suggestion
        return Status::Invalid(
```

pitrou commented on cpp/src/arrow/adapters/orc/adapter.h:
```diff
@@ -272,9 +272,12 @@ class ARROW_EXPORT ORCFileWriter {
+  ///
+  /// \param[in] table the Arrow table from which data is extracted.
   /// \return Status
   Status Write(const Table& table);
```
Do we similarly want to add a `Status Write(const RecordBatch& batch)` method?

LouisClt replied:
Why not, but I guess that an user can still use the Table::FromRecordBatches method to do this.
But it is up to you.

LouisClt replied:
Do you think an implementation of `Status Write(const RecordBatch& batch) `as a method that makes a Table from the RecordBatch and call the already existing method will be ok ?
Else we can do it the other way : transform tables in record batches...

pitrou replied:
The other way would be much more logical IMHO.

wjones127 replied:
I actually think implementing `Status Write(const RecordBatch& batch)` would be more sensible here. The function `ORCFileWriter::Impl::Write()` handles the enforcement of the `batch_size` and (assuming I'm reading correctly) will handle writing batches that straddle chunked array boundaries. If you make `Status Write(const Table& table)` instead implemented by using a `TableBatchReader` and passing each individual batch to the `Write` implementation, then the writer has to break up batches based on the input chunk boundaries. IMO, ideally we should aim to decouple the chunk boundaries in-memory and the serialized chunk boundaries whenever possible.

So unless you want to do a very deep refactoring, I think we should implement `Status Write(const RecordBatch& record_batch)` by just wrapping the batch in a Table.

LouisClt replied:
Thanks for your review. I think that it would be better indeed, to implement `Status Write(const RecordBatch& record_batch)` by wrapping the batch in a table following your reasoning. I will do this, but I will first wait a bit if others want to comment.

pitrou commented on cpp/src/arrow/adapters/orc/adapter_test.cc:
```diff
@@ -239,7 +241,9 @@ void AssertTableWriteReadEqual(const std::shared_ptr<Table>& input_table,
   EXPECT_OK_AND_ASSIGN(auto writer, adapters::orc::ORCFileWriter::Open(
                                         buffer_output_stream.get(), write_options));
-  ARROW_EXPECT_OK(writer->Write(*input_table));
+  for (int i_write = 0; i_write < nb_writes; i_write++) {
```
This is writing N times the same table. Can we instead write a vector of different tables?

LouisClt replied:
Yes I'll try yo do this.

wjones127 commented on cpp/src/arrow/adapters/orc/adapter.cc:
```diff
@@ -781,7 +794,29 @@ Result<std::unique_ptr<ORCFileWriter>> ORCFileWriter::Open(
+    ARROW_ASSIGN_OR_RAISE(auto empty_batch, RecordBatch::MakeEmpty(table.schema()));
+    RETURN_NOT_OK(Write(*empty_batch));
+  } else {
+    TableBatchReader reader(table);
```
(Though see my other comment on whether this approach is the right one)

wjones127 reviewed (commented):
Sorry to drag this out, but when reading more into the implementation it occurred to me that there was good reason that `ORCFileWriter::Impl::Write` took a `Table` and not a `RecordBatch`.

wjones127 reviewed (commented):
This is looking good to me now. Just need to run the formatter.

LouisClt commented:
Good. I was wondering if we need a test for writing a RecordBatch, for the moment this is not tested. Is this ok without test ?

wjones127 commented:
> Good. I was wondering if we need a test for writing a RecordBatch, for the moment this is not tested. Is this ok without test ?

Oh we would prefer to have a test for that.

LouisClt commented:
I added a test for testing the write of RecordBatches with ORC. As the RecodBatch method just makes a call to the "Table" one, I think this is sufficient. The test failures that remains seem to be unrelated.

wjones127 commented:
Sorry I let this sit @LouisClt.

This looks good to me. @pitrou do you have any comments? Or can I merge this?

pitrou commented on cpp/src/arrow/adapters/orc/adapter.cc:
```diff
@@ -727,12 +727,22 @@ class ORCFileWriter::Impl {
+    } else {
+      bool schemas_matching = table.schema()->Equals(arrow_schema_, false);
+      if (!schemas_matching) {
+        return Status::Invalid(
```
Can we return `TypeError` here?

pitrou commented on cpp/src/arrow/adapters/orc/adapter_test.cc:
```diff
@@ -241,7 +241,9 @@ void AssertTableWriteReadEqual(const std::shared_ptr<Table>& input_table,
   EXPECT_OK_AND_ASSIGN(auto writer, adapters::orc::ORCFileWriter::Open(
                                         buffer_output_stream.get(), write_options));
-  ARROW_EXPECT_OK(writer->Write(*input_table));
+  for (auto& input_table : input_tables) {
```
Nit
```suggestion
  for (const auto& input_table : input_tables) {
```

pitrou reviewed (changes requested):
Thanks @LouisClt ! Looks good for the most part, two small comments.
